File: Python/second_largest.py
def sec_largest(arr):

	# # 2n comparisions
	# maxi=arr[0]
	# maxi2=arr[1]
	# for i in range(1,len(arr)):
	# 	if maxi2<arr[i]:	
	# 		maxi2=arr[i]
	# 	if maxi<=arr[i]:
	# 		maxi2=maxi
	# 		maxi=arr[i]
	# return maxi2

	#n+logn-2 comparisions
	compared = find_max_tournament(0,len(arr)-1,arr)
	print compared
	compared2 = find_max_tournament(1,compared[0]-1,compared[1:])
	return compared2[1]

def find_max_tournament(i,j,arr):
	print i,j
	if i==j:
		compared=[]
		compared.append(1)
		compared.append(arr[i])
		return compared

	compared1 = find_max_tournament(i,(i+((j-i)/2)),arr)
	print "compared1",compared1
	compared2 = find_max_tournament(1+(i+((j-i)/2)),j,arr)
	print "compared2",compared2
	# print "before",compared1
	if compared1[1] > compared2[1]:
		k = compared1[0]+1
		# compared1[0] = k
		compared1[0] += 1
		# compared1[k] = compared2[1]
		compared1.append(compared2[1])
		# print "after",compared1
		return compared1

	else:
		k = compared2[0]+1
		# compared2[0] = k
		compared2[0] += 1
		# compared2[k] = compared1[1]
		compared2.append(compared1[1])
		return compared2

if __name__ == '__main__':
	arr = [1,2,3,4,5,6]
	result2 = sec_largest(arr)
	print result2

File: Java/SecondLargest.java
import java.util.Arrays;
import java.util.ArrayList;

public class SecondLargest {
	public static void main(String[] args) {
		int[] arr = {7,4,9,6,5,3};
		System.out.println("Original Array: "+Arrays.toString(arr));

		int result = sec_largest(arr); 
		System.out.println("Second Largest: "+result);
	}

	public static int sec_largest(int[] arr) {
		
		// // 2n comparisions
		// int maxi = arr[0],maxi2 = arr[1];
		// for ( int i=1;i<arr.length;i++ ) {
		// 	if ( maxi2<arr[i] ) {
		// 		maxi2 = arr[i];
		// 	}
		// 	if ( maxi<=arr[i] ) {
		// 		maxi2 = maxi;
		// 		maxi = arr[i];
		// 	}
		// }
		// return maxi2;


		// n+log2-n comparisions
		ArrayList<Integer> compared = find_max_tournament(0,arr.length-1,arr);
		int[] com = new int[compared.size()];
		for (int i = 0; i < com.length; i++) {
    		com[i] = compared.get(i);
		}
		// System.out.println("Original Array: "+Arrays.toString(com));
		ArrayList<Integer> compared2 = find_max_tournament(1,com[0]-1,Arrays.copyOfRange(com,1,com.length));
		return compared2.get(1);
	}

	public static ArrayList<Integer> find_max_tournament(int i,int j,int[] arr) {
		// System.out.println(""+i+","+j);
		if ( i==j ) {
			ArrayList<Integer> compared = new ArrayList<Integer>();
			compared.add(1);
			compared.add(arr[i]);
			return compared;
		}

		ArrayList<Integer> compared1 = find_max_tournament(i,i+((j-i)/2),arr);
		ArrayList<Integer> compared2 = find_max_tournament(1+i+((j-i)/2),j,arr);

		if ( compared1.get(1) > compared2.get(1) ) {
			int k = compared1.get(0)+1;
			compared1.set(0,k);
			compared1.add(compared2.get(1));
			return compared1;
		} else {
			int k = compared2.get(0)+1;
			compared2.set(0,k);
			compared2.add(compared1.get(1));
			return compared2;
		}
	}
}